harvester/harvester
# Enable logging with Harvester upgrades

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**


Upgrade-related logs are hard to keep:

- The logs of the upgrade Jobs are gone almost immediately after the Jobs completed, e.g., pre-drain Jobs
- Existing services like harvester, rancher, and longhorn-manager could have crucial information which might be lost during the upgrade because Pods/containers restart

And support bundles have limitations:

- It can only snapshot a point in time of the cluster; what was lost was lost.

**Solution:**


Introducing UpgradeLog CRD and its accompanying controller collaborate with the existing upgrade procedure to act like a recording assistant.

For detailed information, please refer to the HEP #3274 

**Related Issue:**

#1926

**Test plan:**


1. Prepare a Harvester cluster in master-head version (including this PR)
2. Upgrade with the same ISO
3. Check if the corresponding UpgradeLog created
4. After the DownloadReady condition of the UpgradeLog becomes True, try to generate an archive by executing `curl -k -X POST -u "${CATTLE_ACCESS_KEY}:${CATTLE_SECRET_KEY}" https://<harvester-vip>/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/<upgradelog-name>?action=generate`. Copy the response in stdout (it's the name of the to-be-generated archive)
5. Wait until the archive becomes ready (check this via `upgradelog.status.archives.<archiveName>.ready`), and download it by executing `curl -kfL -u "${CATTLE_ACCESS_KEY}:${CATTLE_SECRET_KEY}" https://<harvester-vip>/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/<upgradelog-name>/download?archiveName=<archiveName> -o <archiveName>.tar.gz`. Check the content of the archive.
6. Users can generate several archives per UpgradeLog and download each of them by doing step 4 and 5 repeatedly during an upgrade
7. After the upgrade is ended (either in success or failure), users can still generate archives and download them. But the content of the archives will be the same because the collection has stopped.
9. Remove the Upgrade resource should purge all the related resources, i.e., the UpgradeLog itself, the ManagedChart (if installed), the Logging resource, the ClusterFlow/Output resources, the log downloader Deployment, the log packager Jobs, and the PVC

Example of commands used to test this PR:

```
# generate an archive
$ curl -k -X POST -u token-jdpkd:k56qxnt28h4nmjp7cpqhzzgkflpkqhbhpppzmkk5bdfmr4p5j2gsn8 \
    https://192.168.48.130/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/hvst-upgrade-lz2q7-upgradelog?action=generate
"hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z" # this is the output

# download the archive
$ curl -fsSL -k -u token-jdpkd:k56qxnt28h4nmjp7cpqhzzgkflpkqhbhpppzmkk5bdfmr4p5j2gsn8 \
    https://192.168.48.130/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/hvst-upgrade-lz2q7-upgradelog/download?archiveName=hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z \
    -o hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z.tar.gz

# check tarball contents
$ tar -tf hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z.tar.gz
./
./logs/
./logs/kube-system.harvester-whereabouts-h6wm4.whereabouts.20230228.log
./logs/kube-system.snapshot-controller-6cc58d858f-9zrvl.snapshot-controller.20230228.log
./logs/kube-system.snapshot-controller-6cc58d858f-8mt8m.snapshot-controller.20230228.log
./logs/kube-system.kube-controller-manager-node-0.kube-controller-manager.20230228.log
./logs/longhorn-system.instance-manager-r-299489e1.replica-manager.20230228.log
./logs/kube-system.rke2-ingress-nginx-controller-5cnzf.rke2-ingress-nginx-controller.20230228.log
./logs/kube-system.kube-apiserver-node-0.kube-apiserver.20230228.log
./logs/cattle-system.rancher-5b8d85d4bc-ksfdg.rancher.20230228.log
./logs/longhorn-system.longhorn-manager-kvnlw.longhorn-manager.20230228.log
./logs/kube-system.rke2-canal-d6lqd.calico-node.20230228.log
./logs/kube-system.etcd-node-0.etcd.20230228.log
./logs/harvester-system.harvester-bcd5f8566-99qr9.apiserver.20230228.log
./logs/cattle-system.apply-hvst-upgrade-lz2q7-prepare-on-node-0-with-4a96a36ff-nfxsx.upgrade.20230228.log
./logs/longhorn-system.instance-manager-e-2db66e15.engine-manager.20230228.log
./logs/harvester-system.hvst-upgrade-lz2q7-apply-manifests-gjbgf.apply.20230228.log
```
